hal/halx86/apic/apictimer.c

```diff
@@ -63,22 +69,75 @@ VOID
 NTAPI
 HalStartProfileInterrupt(IN KPROFILE_SOURCE ProfileSource)
 {
-    UNIMPLEMENTED;
-    return;
+	LVT_REGISTER LvtEntry;
+
+	/* Only handle ProfileTime */
+	if (ProfileSource == ProfileTime)
+	{
+		/* OK, we are profiling now */
+		HalIsProfiling = TRUE;
+
+		/* Set interrupt interval */
+		ApicWrite(APIC_TICR, PCR->HalReserved[0]);
```

Comment:
Slots 0 and 1 already seem to be in use. https://github.com/reactos/reactos/blob/c2c66aff7dacc62d125f2cd61d1167e9a2aa3fd6/hal/halx86/include/halp.h#L49-L50
Should this be using other indices?

In any case, please define some constants for these two entries, as done for HAL_APC_REQUEST/HAL_DPC_REQUEST.

Reply:
> Should this be using other indices?

Again, original Win2003 SP2 x64 kernel uses these. Without testing it's hard to say can it be changed or not. Probably can. But I will #define them for now as is.

Reply:
Ah, it makes sense that those would be different on x64. IRQLs are handled differently there, so the APC/DPC entries are probably not needed. However the APIC HAL also works on x86, so the defines probably need to vary by architecture.

Reply:
IMO, since some entries in HalReserved are BOOLEAN / KIRQL or anything else we should replace this hacky way of using HalReserved entirely and replace it with a structure.
```C
typedef struct HAL_PCR_DATA
{
    ULONG ProfilingInterval;
    ULONG_ProfilingMultiplier;
} HAL_PCR_DATA, *PHAL_PCR_DATA;
C_ASSERT(sizeof(HAL_PCR_DATA) <= FIELD_SIZE(KPCR, HalReserved));

    PHAL_PCR_DATA HalPcrData = (PHAL_PCR_DATA)Pcr->HalReserved;
```